jacal.1
.\" dummy line
.TH JACAL 1 "Feb 1 2008"
.UC 4
.SH NAME
JACAL \- symbolic mathematics system
.SH SYNOPSIS
.B jacal
[ scheme | scm | gsi | gosh | guile | scheme48 | larceny | scmlit | elk | sisc | kawa | mzscheme ]
.br
.sp 0.3
.SH DESCRIPTION
JACAL is an interactive symbolic mathematics program. JACAL can
manipulate and simplify equations, scalars, vectors, matrices, and
tensors of (single and multiple valued) algebraic expressions
containing numbers, variables, radicals, and algebraic and
transcendental functions.
.SH OPTIONS
The optional argument to the
.I jacal
script is the Scheme implementation to run.  Absent the argument, it
searches for implementations in the above order.
.SH ENVIRONMENT VARIABLES
.TP
.B SCHEME_LIBRARY_PATH
is the SLIB Scheme library directory.
.SH FILES
.TP
jacal.info
.br
Texinfo documentation of
.I jacal.
.TP
demo
.br
batch("demo"); to demonstrate commands in
.I jacal
.TP
rw.math
.br
batch("rw.math"); to demonstrate tensors in
"The Robertson-Walker Cosmology Model".
.TP
test.math
.br
batch("test.math"); for regression suite.
.SH AUTHORS
Aubrey Jaffer (agj @ alum.mit.edu)
.br
Michael Thomas
.br
Jerry D. Hedden
.SH SEE ALSO
The JACAL home-page:
.br
http://people.csail.mit.edu/jaffer/JACAL.html
.TP
The full documentation for jacal is maintained as a Texinfo manual. If the info and jacal programs are properly installed at your site, the command
.br
info jacal
.TP
should give you access to the complete manual.


ChangeLog
2026-08-14  Aubrey Jaffer  <agj@alum.mit.edu>

	* jacal.texi (Installation): Added instructions for local install.

	* sexp.scm (saw-newline?): #\nl --> #\newline (for MITScheme).

2026-08-13  Aubrey Jaffer  <agj@alum.mit.edu>

	* Makefile (upgnu): Fighting with gnupload script.

	* jacal.texi, jacal.nsi, Makefile (demo, view.scm): Removed files
	from distribution.

	* jacal.texi (Release Notes, Conventions, Tensors): Updated.

	* init.math, test.math, jacal.texi (lambert_w): Renamed from %w;
	%W is retained.

	* README: Include file manifest.

	* English.scm (one-arg): Warn of extra expressions.

	* sexp.scm (seval): Cleaned up numeric type checking.

2026-08-12  Aubrey Jaffer  <agj@alum.mit.edu>

	* jacal.texi (Tensors, Canonical): Added sections.

	* types.scm (licit:depends): Signal error if passed primitive
	transcendental.

	* rw.math: set width 160; to keep each command on one line.

	* builtin.scm ($): Changed var:pri 9000 --> 65535.
	(%i): Changed var:pri to 2.
	(^): Changed var:pri to 5.

	* init.math (asin, acsc): Changed acot() to atan() of reciprocal.

	* norm.scm (mapply): expr:canonicalize --> normalize argument to
	tcall.  This fixed sin(asin(x)).

	* test.math: Added tests for sec() and csc().

2026-08-11  Aubrey Jaffer  <agj@alum.mit.edu>

	* builtin.scm (%sqrt1, %i): var:pri = 1.  Still problem with %expt
	versus %i; sin(asin(x)).

	* types.scm (poly:pri): Added to calculate var:pri from nesting.
	(var:topo-sort): Removed; not needed with new var:pri ordering.
	(most-nested-arg): Replaces nested-arg.
	(poly:most-nested-var): Replaces poly:one-var.

2026-08-10  Aubrey Jaffer  <agj@alum.mit.edu>

	* jacal.texi (Defining Differential Equations): Added memoized
	recurrence definitions and factorial example.

	* test.math (Transcendental reductions): added tests.

	* builtin.scm (listofvars)
	* types.scm (licit:depends)
	* sexp.scm (batch1)
	* poly.scm (poly:vars, licit:vars)
	* norm.scm (total-chain-exts)
	* hensel.scm (make-random-ideal)
	* ext.scm (alg:vars)
	* elim.scm (capply, var:max-lambda-position)
	(var:min-lambda-position): Change math:equal? procedures to eqv?
	analogs when dealing with vars or integers.

	* types.scm (sexp->new-var): Nesting calls increases priority.
	(var:topo-sort): Abstracted from licit:depends.

	* builtin.scm (differential): Just calls total-differential.
	(priority): "show priority" now prints the priorities of all vars.

	* ext.scm (extensions, chainables): Greatly simplified using
	licit:depends.

	* debug.scm (math:debug): set true.

	* grammar.scm (math:advise): Added; prints on when math:debug is on.

	* types.scm (licit:depends): Changed to topological-sort.
	(nested-arg): Changed to math:advise 'multiple-nested-vars.

	* elim.scm (eliminate): In cases of singular-reduction,
	math:advise instead of math:error; and return novalue.
	(ext:elim): (reduce math:union (map extensions poleqns)) -->
	(extensions poleqns).

	* sexp.scm (math:greet): help() --> help;.

2026-08-09  Aubrey Jaffer  <agj@alum.mit.edu>

	* types.scm (licit:depends): Scans var:arglist for depends.
	(register-instance!): does this twice!
	(var:set-depends! var (remove var (licit:depends impl)))

	* ext.scm (var:application?): Renamed from "application?"
	(extends): Try simplification.

	* sexp.scm (seval): Disabled expected-abstract-transcendental-var-
	clause; not sure what it was for.

	* types.scm (math:adjoin): Simplified.
	(var:reset!, var:definverse!): Changed (var:set-func! var #f) to
	(var:set-func! var (var->expl var))

	* builtin.scm (func): Simplified.

2026-08-08  Aubrey Jaffer  <agj@alum.mit.edu>

	* jacal.texi (Algebraic Operators): Recurrences will eventually be
	automatically generated.

	* norm.scm (total-differential): Disabled equation as argument.

	* types.scm (var:reset!, var:definverse!): Corrected setting of
	arglist and depends.

	* ext.scm (chainables): Defined in terms of extensions().

	* builtin.scm (extrule): extrule --> definition.

	* jacal.texi (Algebraic Commands): extrule --> definition.

	* types.scm (poleqn:differential?, var:nodiffs): Changed to use
	var:differential?.

	* English.scm (tps:tex): Added %W and %w.

2026-08-07  Aubrey Jaffer  <agj@alum.mit.edu>

	* jacal.texi (Miscellaneous, Running Jacal, Algebraic Commands):
	improved text.

	* types.scm (with-shadows): removed.

	* builtin.scm (depvars cartprod): removed.
	(extrule): Now returns an equation in all 3 cases.

	* English.scm (tps:2d, tps:c, tps:standard, tps:tex)
	* std.scm (tps:std): Removed qed and help.

	* builtin.scm (factorial): Now only accepts integer argument.

	* English.scm (tps:2d, tps:c, tps:standard, tps:tex)
	* std.scm (tps:std)
	* unparse.scm (hformat, unparse): Moved rules for symbol lookup
	into a list at start of template; Moved template: rules into list
	in cadr of template; the remaining (function) rules are in cddr of
	template.  This mitigates the speed impact of doing symbol lookup.

	* English.scm (tps:2d, tps:standard, tps:tex): Removed % as command.
	(tps:tex): Added atan(), tan(), exp(), and log() functions.
	Added translation for %i.

	* builtin.scm (transcript-files): Added so transcript-off isn't
	called with no active transcript.

	* sexp.scm (batch1): When no *ingrammar* is set, echo with 'standard.

	* init.math (*PQ): Moved after function definitions.

	* jacal.texi (Flags): Removed @ from TeX example.

	* norm.scm (import-var): Don't choke if not defined yet.

2026-08-06  Aubrey Jaffer  <agj@alum.mit.edu>

	* jacal.nsi: missing files added.

	* jacal.texi (Running Jacal): Expanded.
	(Miscellaneous): Reran examples.

	* jacal.nsi: Removed "DOC\history".

	* builtin.scm (flag-set, flag-get): Fixed for "priority" flag.

	* Makefile (jacal.doc): Removed 'ul' command.

2026-08-05  Aubrey Jaffer  <agj@alum.mit.edu>

	* toploads.scm (*jacal-version*): Bumped from 1c9 to 2a1.

2026-08-04  Aubrey Jaffer  <agj@alum.mit.edu>

	* jacal.texi (Transcendental Functions): Chapter added.


jacal.sh

##"jacal" sh script; Find a Scheme implementation and run JACAL in it.
# Copyright (C) 2001, 2003, 2004 Aubrey Jaffer
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or (at
# your option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

usage="Usage: jacal [--version | -v]

  Display version information and exit successfully.

Usage: jacal SCHEME

  Start JACAL using Scheme implementation SCHEME.

Usage: jacal

  Run JACAL using (MIT)'scheme', 'scm', 'gsi', 'gosh', 'guile',
  'slib48', 'larceny', 'scmlit', 'elk', 'sisc', 'kawa', or
  'mzscheme'."

case "$1" in
    -v | --ver*) echo jacal "$VERSION"; exit 0;;
    "") if type scheme>/dev/null 2>&1; then
	command=scheme
	fi;;
    -*) echo "$usage"; exit 1;;
    *) command="$1"
	shift
esac

if [ -z "$command" ]; then
    if type scm>/dev/null 2>&1; then
	command=scm; implementation=scm
    elif type gsi>/dev/null 2>&1; then
	command=gsi; implementation=gam
    elif type gosh>/dev/null 2>&1; then
	command=gosh; implementation=gch
    elif type guile>/dev/null 2>&1; then
	command=guile; implementation=gui
    elif type slib48>/dev/null 2>&1; then
	echo "do 'cd ${JACALDIR}; make jacal48'"
    elif type scheme48>/dev/null 2>&1; then
	echo "do 'cd ${SCHEME_LIBRARY_PATH}; make slib48'"
	echo "do 'cd ${JACALDIR}; make jacal48'"
    elif type larceny>/dev/null 2>&1; then
	command=larceny; implementation=lar
    elif type scmlit>/dev/null 2>&1; then
	command=scmlit; implementation=scm
    elif type elk>/dev/null 2>&1; then
	command=elk; implementation=elk
    elif type sisc>/dev/null 2>&1; then
	command=sisc; implementation=ssc
    elif type kawa>/dev/null 2>&1; then
	command=kawa; implementation=kwa
    elif type mzscheme>/dev/null 2>&1; then
	command=mzscheme; implementation=plt
    else
	echo No Scheme implementation found.
	exit 1
    fi
# Gambit 4.0 doesn't allow input redirection; foils --version test.
elif [ "$command" = "gsi" ]; then implementation=gam
elif type $command>/dev/null 2>&1; then
    SPEW="`$command --version < /dev/null 2>&1`"
    if   echo ${SPEW} | grep 'Initialize load-path (colon-list of directories)'\
				       >/dev/null 2>&1; then implementation=elk
    elif echo ${SPEW} | grep 'MIT'       >/dev/null 2>&1; then implementation=mit
    elif echo ${SPEW} | grep 'UMB Scheme'>/dev/null 2>&1; then implementation=umb
    elif echo ${SPEW} | grep 'scheme48'  >/dev/null 2>&1; then implementation=s48
    elif echo ${SPEW} | grep 'larceny'   >/dev/null 2>&1; then implementation=lar
    elif echo ${SPEW} | grep 'Guile'     >/dev/null 2>&1; then implementation=gui
    elif echo ${SPEW} | grep 'gosh'      >/dev/null 2>&1; then implementation=gch
    elif echo ${SPEW} | grep 'SCM'       >/dev/null 2>&1; then implementation=scm
    elif echo ${SPEW} | grep 'SISC'      >/dev/null 2>&1; then implementation=ssc
    elif echo ${SPEW} | grep 'Kawa'      >/dev/null 2>&1; then implementation=kwa
    elif echo ${SPEW} | grep 'MzScheme'  >/dev/null 2>&1; then implementation=plt
    else implementation=
    fi
else
    echo "Program '$command' not found."
    exit 1
fi

case $implementation in
    scm);;
    s48);;
    *) if [ -z "${SCHEME_LIBRARY_PATH}" ]; then
	if type rpm>/dev/null 2>&1; then
	    SCHEME_LIBRARY_PATH=`rpm -ql slib 2>/dev/null \
		| grep require.scm | sed 's%require.scm%%'`
	fi
	fi
	if [ -z "${SCHEME_LIBRARY_PATH}" ]; then
	    if [ -d /usr/local/lib/slib/ ]; then
		SCHEME_LIBRARY_PATH=/usr/local/lib/slib/
	    elif [ -d /usr/share/slib/ ]; then
		SCHEME_LIBRARY_PATH=/usr/share/slib/
	    fi
	    export SCHEME_LIBRARY_PATH
	fi;;
esac

# for gambit
case $implementation in
    gam) if [ -z "${LD_LIBRARY_PATH}" ]; then
	LD_LIBRARY_PATH=/usr/local/lib
	export LD_LIBRARY_PATH
	fi;;
esac

case $implementation in
    scm) echo $command -ip1 -l ${JACALDIR}go "$@"
	exec $command -ip1 -l ${JACALDIR}go "$@";;
    elk) echo $command -i -l ${JACALDIR}elk.scm "$@"
	exec $command -i -l ${JACALDIR}elk.scm "$@";;
    gam) echo $command -:s ${SCHEME_LIBRARY_PATH}gambit.init ${JACALDIR}go.scm "$@"
	exec $command -:s ${SCHEME_LIBRARY_PATH}gambit.init ${JACALDIR}go.scm "$@";;
    gch) echo $command -l${SCHEME_LIBRARY_PATH}gosh.init -l${JACALDIR}go "$@"
	exec $command -l${SCHEME_LIBRARY_PATH}gosh.init -l${JACALDIR}go "$@";;
    gui) if [ -f ${SCHEME_LIBRARY_PATH}guile.use ]; then
	echo $command -l ${SCHEME_LIBRARY_PATH}guile.init -l ${SCHEME_LIBRARY_PATH}guile.use -l ${JACALDIR}go.scm "$@"
	exec $command -l ${SCHEME_LIBRARY_PATH}guile.init -l ${SCHEME_LIBRARY_PATH}guile.use -l ${JACALDIR}go.scm "$@"
	else
	echo $command -l ${SCHEME_LIBRARY_PATH}guile.init -l ${JACALDIR}go.scm "$@"
	exec $command -l ${SCHEME_LIBRARY_PATH}guile.init -l ${JACALDIR}go.scm "$@"
	fi;;
    ssc) echo $command -e "(begin (load \"${SCHEME_LIBRARY_PATH}sisc.init\") (load \"${JACALDIR}go.scm\"))" -- "$@"
	exec $command -e "(begin (load \"${SCHEME_LIBRARY_PATH}sisc.init\") (load \"${JACALDIR}go.scm\"))" -- "$@";;
    kwa) echo $command -f ${SCHEME_LIBRARY_PATH}kawa.init -f ${JACALDIR}go.scm -- "$@"
	exec $command -f ${SCHEME_LIBRARY_PATH}kawa.init -f ${JACALDIR}go.scm -- "$@";;
    lar) echo $command -- -e "(require 'srfi-96)" ${JACALDIR}go.scm "$@"
	exec $command -- -e "(require 'srfi-96)" ${JACALDIR}go.scm "$@";;
    mit) echo $command -load ${SCHEME_LIBRARY_PATH}mitscheme.init -load ${JACALDIR}go "$@"
	exec $command -load ${SCHEME_LIBRARY_PATH}mitscheme.init -load ${JACALDIR}go "$@";;
    plt) echo $command -f ${SCHEME_LIBRARY_PATH}mzscheme.init -f ${JACALDIR}go.scm "$@"
	exec $command -f ${SCHEME_LIBRARY_PATH}mzscheme.init -f ${JACALDIR}go.scm "$@";;
    umb) echo umb-scheme does not run jacal; exit 1;;
    s48) if [ -f "${JACALDIR}scheme48.image" ]; then
	echo ";;; Type (math); to begin."
	echo scheme48 -h 4500000 -i ${JACALDIR}scheme48.image "$@"
	exec scheme48 -h 4500000 -i ${JACALDIR}scheme48.image "$@"
	else
	echo "scheme48 found; do: 'cd ${JACALDIR}; make jacal48'";
	fi
	exit 1;;
    *)   echo Program '$command' ??
	exit 1;;
esac
